native/cocos/renderer/gfx-vulkan/VKPipelineCache.cpp

```diff
@@ -0,0 +1,120 @@
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+****************************************************************************/
+
+#include "VKPipelineCache.h"
+#include "gfx-base/GFXUtil.h"
+#include "base/BinaryArchive.h"
+#include <fstream>
+
+namespace cc::gfx {
+
+static const char* fileName = "/pipeline_cache_vk.bin";
+
+static const uint32_t MAGIC   = 0x4343564B; // "CCVK"
+static const uint32_t VERSION = 1;
+
+static void loadData(ccstd::vector<char> &data) {
+    auto path = getPipelineCacheFolder() + fileName;
+    std::ifstream stream(path, std::ios::binary);
+    if (!stream.is_open()) {
+        CC_LOG_INFO("Load program cache, no cached files.");
+        return;
+    }
+
+    uint32_t magic = 0;
+    uint32_t version = 0;
+
+    BinaryInputArchive archive(stream);
+    auto loadResult = archive.load(magic);
+    loadResult &= archive.load(version);
+
+    uint32_t size = 0;
+    if (loadResult && magic == MAGIC && version >= VERSION) {
+        loadResult &= archive.load(size);
+        data.resize(size);
+        loadResult &= archive.load(data.data(), size);
+    }
+    if (loadResult) {
+        CC_LOG_INFO("Load pipeline cache success.");
+    }
+}
+
+CCVKPipelineCache::~CCVKPipelineCache() {
+    if (_pipelineCache != VK_NULL_HANDLE) {
+#if CC_USE_PIPELINE_CACHE
+        saveCache();
+#endif
+        vkDestroyPipelineCache(_device, _pipelineCache, nullptr);
+    }
+}
+
+void CCVKPipelineCache::init(VkDevice dev) {
+    _device = dev;
+    loadCache();
+}
+
+void CCVKPipelineCache::loadCache() {
+    ccstd::vector<char> data;
+#if CC_USE_PIPELINE_CACHE
+    loadData(data);
+#endif
+
+    VkPipelineCacheCreateInfo cacheInfo = {};
+    cacheInfo.sType = VK_STRUCTURE_TYPE_PIPELINE_CACHE_CREATE_INFO;
+    cacheInfo.pNext = nullptr;
+    cacheInfo.initialDataSize = static_cast<uint32_t>(data.size());
+    cacheInfo.pInitialData = data.data();
+    VK_CHECK(vkCreatePipelineCache(_device, &cacheInfo, nullptr, &_pipelineCache));
+}
+
+void CCVKPipelineCache::saveCache() {
+    if (!_dirty) {
+        return;
+    }
+    auto path = getPipelineCacheFolder() + fileName;
+    std::ofstream stream(path, std::ios::binary);
+    if (!stream.is_open()) {
+        CC_LOG_INFO("Save program cache failed.");
+        return;
+    }
+    BinaryOutputArchive archive(stream);
+    archive.save(MAGIC);
+    archive.save(VERSION);
+
+    size_t size = 0;
+    vkGetPipelineCacheData(_device, _pipelineCache, &size, nullptr);
+    ccstd::vector<char> data(size);
+    vkGetPipelineCacheData(_device, _pipelineCache, &size, data.data());
+
+    archive.save(static_cast<uint32_t>(size));
+    archive.save(data.data(), static_cast<uint32_t>(size));
+    _dirty = false;
+}
+
+VkPipelineCache CCVKPipelineCache::getHandle() {
+    _dirty = true;
```

Comment:
It is wired to set dirty here.

Reply:
I have split it into two functions: `setDirty` and `getHandle`.